share/sysbuild/cmake/modules/sysbuild_extensions.cmake

```diff
@@ -134,6 +134,8 @@ function(ExternalZephyrProject_Add)
     )
   endif()
 
+  set_property(GLOBAL APPEND PROPERTY sysbuild_images ${ZBUILD_APPLICATION})
```

Comment:
This should not be added here.

The existing code at those places:
https://github.com/zephyrproject-rtos/zephyr/blob/7d898edfc5c2e8dfa4bc158c5f74614f15dd833b/share/sysbuild/CMakeLists.txt#L88
https://github.com/zephyrproject-rtos/zephyr/blob/7d898edfc5c2e8dfa4bc158c5f74614f15dd833b/share/sysbuild/bootloader/CMakeLists.txt#L13

are exactly done outside of `ExternalZephyrProject_Add()` in order to keep building of Zephyr based projects independent of flashing.
A use-case is where a sysbuild module / extension want to build a Zephyr based project but merge the output of such project into another image.

Also, not all images may be Zephyr project. There can be ather forms of external projects, which can be added to sysbuild image list. Of course flashing of such images requires that they generate a runners.yml file, but such file is quite simple.
Such a use-case could be porting of the espressif bootloader to sysbuild, ref:
https://github.com/zephyrproject-rtos/zephyr/blob/7d898edfc5c2e8dfa4bc158c5f74614f15dd833b/soc/xtensa/esp32s3/CMakeLists.txt#L29-L30

So please keep the list of sysbuild images to flash independent of the `ExternalZephyrProject_Add()`.

Of course this comment naturally also impact the later GROUP argument proposal, cause if `ExternalZephyrProject_Add()` is decoupled from sysbuild image list, then it should also be decoupled from sysbuild image groups.

Reply:
> The existing code at those places (...) are exactly done outside of `ExternalZephyrProject_Add()` in order to keep building of Zephyr based projects independent of flashing. A use-case is where a sysbuild module / extension want to build a Zephyr based project but merge the output of such project into another image (instead of flashing said Zephyr image directly).

You mean to support building an image without flashing it? I was not aware of this and I'm pretty sure this is undocumented. I would propose specifying that as an argument to `ExternalZephyrProject_Add()` as well, so that the `IMAGES` variable can be removed.

> Also, not all images may be Zephyr project. There can be other forms of external projects which can be added to sysbuild image list.

My impression was that non-Zephyr projects were not really closely tied to sysbuild. Here's the documentation:

https://github.com/zephyrproject-rtos/zephyr/blob/2f718dd369e41a9ae50fb613ac29d8ce192da440/doc/build/sysbuild/index.rst?plain=1#L622-L635

I wouldn't presume it necessary to populate `IMAGES` when using standard `ExternalProject_Add()`, because I could call it with `BUILD_ALWAYS TRUE` to ensure that it builds. For Zephyr projects, sysbuild will ignore them unless I append to `IMAGES`. Also, I'm not sure if there are any side effects of calling `ExternalZephyrProject_Cmake()` on a target not added using `ExternalZephyrProject_Add()`. All of this should probably be documented.

By the way, is there a reason why the espressif bootloader won't be able to use `ExternalZephyrProject_Add()`?

Reply:
> You mean to support building an image without flashing it?

Correct.

> I was not aware of this and I'm pretty sure this is undocumented. 

It's been discussed as part of a potential wish for systems to be able to build Zephyr samples, but decide to merge to single / fewer combined images instead of flashing each image independently. That would currently be a very advanced use, but is the reason why building and flashing is not intermixed, as that could make it harder to split / make flashing more independent of building later. 
Also, it's not directly described that `ExternalZephyrProject_Add()` will also do flashing: 
https://github.com/zephyrproject-rtos/zephyr/blob/7e23e73bc186da4192a75f367e8d4f53441016de/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L112-L113

But it's a very valid point that we should improve the doc details on this.
Especially after the split of adding the project and running CMake in this commit: https://github.com/zephyrproject-rtos/zephyr/commit/631fa636100b9605dfbc9bfb1968fdfae656ddd4 it is neccesary to ensure CMake is still executed, which is also done based on the `IMAGES` variable.

> I would propose specifying that as an argument to `ExternalZephyrProject_Add()` as well, so that the `IMAGES` variable can be removed.

Need some careful consideration. Adding an argument to `ExternalZephyrProject_Add()` would indicate it's the caller of the function who decides this, where in some cases it could actually be a downstream project who would like to control this.
It's a very valid discussion you're opening, and perhaps it's time to do some more consideration on this topic.

Reply:
> By the way, is there a reason why the espressif bootloader won't be able to use ExternalZephyrProject_Add()?

The fact that the espressif bootloader is not a Zephyr project 😉

Reply:
> Also, I'm not sure if there are any side effects of calling `ExternalZephyrProject_Cmake()` on a target not added using `ExternalZephyrProject_Add()`.

Handled here:
https://github.com/zephyrproject-rtos/zephyr/blob/215f180296c9148a468eb294090e0209629cc81f/share/sysbuild/cmake/modules/sysbuild_extensions.cmake#L298-L302

Reply:
> The fact that the espressif bootloader is not a Zephyr project wink

I don't believe that this is true anymore, I think it has been switched to mcuboot

Reply:
> I don't believe that this is true anymore, I think it has been switched to mcuboot

both MCUboot and Espressif IDF bootloaders are supported as 2nd stage bootloader for esp32.

And the support of MCUboot on esp32 still doesn't make the IDF bootloader a Zephyr project 😉